File: wear/src/main/java/fr/xgouchet/android/backintime/renderer/GroundHogRenderer.java
package fr.xgouchet.android.backintime.renderer;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Typeface;
import android.graphics.drawable.BitmapDrawable;

import java.util.Calendar;
import java.util.Locale;

import fr.xgouchet.android.backintime.R;

public class GroundHogRenderer extends AbstractRenderer {


    private static String FONT_PATH = "groundhog.ttf";

    private final Bitmap mBitmap;
    private final Canvas mCanvas;
    private BitmapDrawable mBackground;
    private Bitmap mHighlights, mShadows, mShadowsDimmed;

    private int mWidth, mHeight;
    private int mTextHourX, mTextMinX, mTextY;
    private float mTextSize, mTextWidth;

    private Typeface mTypeface;
    private Paint mWhiteTextPaint, mGreyTextPaint, mBlackPaint, mOverlay, mMultiply;

    public GroundHogRenderer(Context context) {

        //
        Resources res = context.getResources();

        // load everything
        loadSizes(res);
        loadFont(context);
        loadPaints(res);

        // create main bitmap
        mBitmap = Bitmap.createBitmap(mWidth, mHeight, Bitmap.Config.ARGB_4444);
        mCanvas = new Canvas(mBitmap);

        // create compositing bitmaps
        mHighlights = Bitmap.createBitmap(mWidth, mHeight, Bitmap.Config.ARGB_4444);
        mShadows = Bitmap.createBitmap(mWidth, mHeight, Bitmap.Config.ARGB_4444);
        mShadowsDimmed = Bitmap.createBitmap(mWidth, mHeight, Bitmap.Config.ARGB_4444);
        loadImages(res);
    }


    @Override
    public Bitmap render(Context context, Calendar time, boolean dimmed) {

        // draw background
        if (dimmed) {
            mCanvas.drawRect(0, 0, mWidth, mHeight, mBlackPaint);
        } else {
            mBackground.draw(mCanvas);
        }

        // draw text
        Paint textPaint = dimmed ? mWhiteTextPaint : mGreyTextPaint;
        drawTextCentered(String.format(Locale.US, "%02d", time.get(Calendar.HOUR_OF_DAY)), mTextHourX, mTextY, textPaint);
        drawTextCentered(String.format(Locale.US, "%02d", time.get(Calendar.MINUTE)), mTextMinX, mTextY, textPaint);

        if (dimmed) {
            // only draw shadow
            mCanvas.drawBitmap(mShadowsDimmed, 0, 0, mMultiply);
        } else {
            // draw highlight
            mCanvas.drawBitmap(mHighlights, 0, 0, mOverlay);
            // draw shadow
            mCanvas.drawBitmap(mShadows, 0, 0, mMultiply);
        }

        return mBitmap;
    }

    private void drawTextCentered(String text, int x, int y, Paint paint) {
        float measuredWidth = paint.measureText(text);
        float oversize = (mTextWidth - measuredWidth) / 2;

        mCanvas.drawText(text, x + oversize, y, paint);
    }

    private void loadPaints(Resources res) {
        mWhiteTextPaint = new Paint();
        mWhiteTextPaint.setAntiAlias(false);
        mWhiteTextPaint.setSubpixelText(false);
        mWhiteTextPaint.setTypeface(mTypeface);
        mWhiteTextPaint.setStyle(Paint.Style.FILL);
        mWhiteTextPaint.setColor(Color.WHITE);
        mWhiteTextPaint.setTextSize(mTextSize);

        mGreyTextPaint = new Paint();
        mGreyTextPaint.setAntiAlias(true);
        mGreyTextPaint.setSubpixelText(false);
        mGreyTextPaint.setTypeface(mTypeface);
        mGreyTextPaint.setStyle(Paint.Style.FILL);
        mGreyTextPaint.setColor(res.getColor(R.color.groundhog_text_color));
        mGreyTextPaint.setTextSize(mTextSize);


        mBlackPaint = new Paint();
        mBlackPaint.setColor(Color.BLACK);
        mBlackPaint.setStyle(Paint.Style.FILL);

        mOverlay = new Paint();
        mOverlay.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.OVERLAY));

        mMultiply = new Paint();
        mMultiply.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.MULTIPLY));
    }

    private void loadFont(Context context) {
        mTypeface = Typeface.createFromAsset(context.getAssets(), FONT_PATH);
    }

    private void loadImages(Resources res) {
        // load images
        mBackground = (BitmapDrawable) res.getDrawable(R.drawable.groundhog_background);
        mBackground.setBounds(0, 0, mWidth, mHeight);

        // load images into bitmaps
        BitmapDrawable temp = (BitmapDrawable) res.getDrawable(R.drawable.groundhog_highlights);
        temp.setBounds(0, 0, mWidth, mHeight);
        temp.draw(new Canvas(mHighlights));

        temp = (BitmapDrawable) res.getDrawable(R.drawable.groundhog_shadows);
        temp.setBounds(0, 0, mWidth, mHeight);
        temp.draw(new Canvas(mShadows));

        temp = (BitmapDrawable) res.getDrawable(R.drawable.groundhog_shadows_dimmed);
        temp.setBounds(0, 0, mWidth, mHeight);
        temp.draw(new Canvas(mShadowsDimmed));
    }

    private void loadSizes(Resources res) {

        // get the image size
        mWidth = res.getDimensionPixelSize(R.dimen.groundhog_width);
        mHeight = res.getDimensionPixelSize(R.dimen.groundhog_height);

        // load the text size
        mTextSize = res.getDimensionPixelSize(R.dimen.groundhog_text_size);
        mTextWidth = res.getDimensionPixelSize(R.dimen.groundhog_text_width);

        // load the text position
        mTextHourX = res.getDimensionPixelOffset(R.dimen.groundhog_text_hour_x);
        mTextMinX = res.getDimensionPixelOffset(R.dimen.groundhog_text_min_x);
        mTextY = res.getDimensionPixelOffset(R.dimen.groundhog_text_y);
    }


}

File: backintime-wear/src/main/java/fr/xgouchet/backintime/wear/GroundhogRenderer.kt
package fr.xgouchet.backintime.wear

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Typeface
import android.graphics.drawable.BitmapDrawable
import java.util.Calendar
import java.util.Locale
import kotlin.math.min

class GroundhogRenderer(context: Context) {
    companion object {
        const val FONT_PATH = "groundhog.ttf"
    }


    private var width: Int = 0
    private var height: Int = 0

    private var background: BitmapDrawable? = null
    private var highlights: Bitmap? = null
    private var shadows: Bitmap? = null
    private var shadowsDimmed: Bitmap? = null

    private var digitsTextSize: Int = 0
    private var textWidth: Int = 0
    private var textHourX: Int = 0
    private var textMinX: Int = 0
    private var textY: Int = 0
    private var offsetScale: Float = 0f

    private var digitsTypeface: Typeface? = null

    private var whiteTextPaint: Paint = Paint()
    private var greyTextPaint: Paint = Paint()
    private var blackPaint: Paint = Paint()
    private var overlayPaint: Paint = Paint()
    private var multiplyPaint: Paint = Paint()

    init {
        val res = context.resources

        // load everything
        loadSizes(res)
        loadFont(context)
        loadPaints(res)


        highlights = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_4444)
        shadows = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_4444)
        shadowsDimmed = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_4444)
        loadDrawables(res)
    }


    // region Loading

    private fun loadSizes(res: Resources) {

        // get sizes
        width = res.getDimensionPixelSize(R.dimen.groundhog_width)
        height = res.getDimensionPixelSize(R.dimen.groundhog_height)

        // load the text size
        digitsTextSize = res.getDimensionPixelSize(R.dimen.groundhog_text_size)
        textWidth = res.getDimensionPixelOffset(R.dimen.groundhog_text_width)

        // load the text position
        textHourX = res.getDimensionPixelOffset(R.dimen.groundhog_text_hour_x)
        textMinX = res.getDimensionPixelOffset(R.dimen.groundhog_text_min_x)
        textY = res.getDimensionPixelOffset(R.dimen.groundhog_text_y)

        offsetScale = res.getDimensionPixelOffset(R.dimen.groundhog_offset_scale).toFloat()
    }

    private fun loadDrawables(res: Resources) {
        // load images
        background = res.getDrawable(R.drawable.groundhog_background).apply {
            setBounds(0, 0, width, height)
        } as BitmapDrawable

        // load images into bitmaps
        res.getDrawable(R.drawable.groundhog_highlights).apply {
            setBounds(0, 0, width, height)
            draw(Canvas(highlights))
        }

        res.getDrawable(R.drawable.groundhog_shadows).apply {
            setBounds(0, 0, width, height)
            draw(Canvas(shadows))
        }

        res.getDrawable(R.drawable.groundhog_shadows_dimmed).apply {
            setBounds(0, 0, width, height)
            draw(Canvas(shadowsDimmed))
        }
    }

    private fun loadFont(context: Context) {
        digitsTypeface = Typeface.createFromAsset(context.assets, FONT_PATH)
    }

    private fun loadPaints(res: Resources) {

        whiteTextPaint = Paint().apply {
            isAntiAlias = false
            isSubpixelText = false
            typeface = digitsTypeface
            style = Paint.Style.FILL
            color = Color.WHITE
            textSize = digitsTextSize.toFloat()
        }

        greyTextPaint = Paint().apply {
            isAntiAlias = true
            isSubpixelText = false
            typeface = digitsTypeface
            style = Paint.Style.FILL
            color = res.getColor(R.color.groundhog_text_color)
            textSize = digitsTextSize.toFloat()
        }

        blackPaint = Paint().apply {
            color = Color.BLACK
            style = Paint.Style.FILL
        }

        overlayPaint = Paint().apply {
            xfermode = PorterDuffXfermode(PorterDuff.Mode.OVERLAY)
        }

        multiplyPaint = Paint().apply {
            xfermode = PorterDuffXfermode(PorterDuff.Mode.MULTIPLY)
        }

    }

    // endregion

    // region Rendering

    fun render(canvas: Canvas, calendar: Calendar, dimmed: Boolean, lowBitAmbient: Boolean) {

        whiteTextPaint.isAntiAlias = !lowBitAmbient

        canvas.save()

        val scale = min(canvas.width.toFloat() / width.toFloat(), canvas.height.toFloat() / height.toFloat())
        val offsetX = ((canvas.width / scale) - width.toFloat()) / 2f
        val offsetY = ((canvas.height / scale) - height.toFloat()) / 2f

        canvas.scale(scale, scale)
        canvas.translate(offsetX, offsetY)


        drawBackground(canvas, dimmed)
        drawTime(canvas, calendar, dimmed)

        drawHighlightsAndShadows(dimmed, canvas)

        canvas.restore()
    }

    private fun drawHighlightsAndShadows(dimmed: Boolean, canvas: Canvas) {
        if (dimmed) {
            // only draw shadow
            canvas.drawBitmap(shadowsDimmed, 0f, 0f, multiplyPaint)
        } else {
            // draw highlight
            canvas.drawBitmap(highlights, 0f, 0f, overlayPaint)
            // draw shadow
            canvas.drawBitmap(shadows, 0f, 0f, multiplyPaint)
        }
    }

    private fun drawBackground(canvas: Canvas, dimmed: Boolean) {
        if (dimmed) {
            canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), blackPaint)
        } else {
            background?.draw(canvas)
        }
    }

    private fun drawTime(canvas: Canvas, calendar: Calendar, dimmed: Boolean) {
        // draw text
        val textPaint = if (dimmed) whiteTextPaint else greyTextPaint
        drawTextCentered(canvas, String.format(Locale.US, "%d", calendar.get(Calendar.HOUR_OF_DAY)), textHourX, textY, textPaint)
        drawTextCentered(canvas, String.format(Locale.US, "%02d", calendar.get(Calendar.MINUTE)), textMinX, textY, textPaint)
    }

    private fun drawTextCentered(canvas: Canvas, text: String, x: Int, y: Int, paint: Paint) {
        val measuredWidth = paint.measureText(text)
        val oversize = (textWidth - measuredWidth) / 2

        canvas.save()

        canvas.clipRect(0, 0, width, height / 2)
        canvas.drawText(text, x + oversize, y - offsetScale, paint)

        canvas.restore()

        canvas.save()

        canvas.clipRect(0, height / 2, width, height)
        canvas.drawText(text, x + oversize + offsetScale, y.toFloat(), paint)

        canvas.restore()

    }

    // endregion
}